crates/ruma-events/src/room/message.rs

```diff
@@ -696,13 +696,26 @@ impl TextMessageEventContent {
     }
 
     /// A convenience constructor to create a markdown message.
+    ///
+    /// Returns an html message if some markdown formatting was detected, otherwise returns a plain
+    /// message.
     #[cfg(feature = "markdown")]
     #[cfg_attr(docsrs, doc(cfg(feature = "markdown")))]
     pub fn markdown(body: impl Into<String>) -> Self {
+        use pulldown_cmark::{html::push_html, Event, Parser, Tag};
+
         let body = body.into();
-        let mut html_body = String::new();
-        pulldown_cmark::html::push_html(&mut html_body, pulldown_cmark::Parser::new(&body));
-        Self::html(body, html_body)
+
+        if Parser::new(&body).all(|e| {
```

Comment:
This means we're parsing twice, no? Can't we bind this to a variable and borrow?

```rust
let parsed = Parser::new(&body);

if parsed.all(|e| {
    ....
}) {
...
} else {
....
}
```

Another thing, do we want to use `.all()` here? I think this is what element might be doings but catching the common case where it's just one paragraph enclosing a single text element might be the better idea.

Are we sure that multiple paragraphs aren't desirable as html and are we sure that we want to go through all parsed elements here?

Reply:
The idea was also to save data so I thought parsing all paragraphs made sense, but you're probably right that separate paragraphs should be formatted, since paragraphs are some kind of formatting. And it'll be probably more consistent visually between messages with formatting and messages without formatting.